hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/HazelcastWriters.java

```diff
@@ -189,6 +328,145 @@ public String toString() {
         }
     }
 
+    public static class EntryProcessorWriter<T, K, V> extends AbstractProcessor {
+
+        private static final int MAX_PARALLEL_ASYNC_OPS = 1000;
+        private final AtomicInteger numConcurrentOps = new AtomicInteger();
+
+        private final IMap<K, V> map;
+        private final DistributedFunction<T, K> toKeyFn;
+        private final DistributedFunction<T, EntryProcessor<K, V>> toEntryProcessorFn;
+        private final AtomicReference<Throwable> lastError = new AtomicReference<>();
+        private final HazelcastInstance instance;
+        private final ExecutionCallback callback = callbackOf(
+                response -> numConcurrentOps.decrementAndGet(),
+                exception -> {
+                    numConcurrentOps.decrementAndGet();
+                    if (exception != null) {
+                        lastError.compareAndSet(null, exception);
+                    }
+                });
+        private final boolean isLocal;
+
+
+        public EntryProcessorWriter(HazelcastInstance instance, String name,
+                                    DistributedFunction toKeyFn,
+                                    DistributedFunction<T, EntryProcessor<K, V>> toEntryProcessorFn,
+                                    boolean isLocal) {
+            this.instance = instance;
+            this.map = instance.getMap(name);
+            this.toKeyFn = toKeyFn;
+            this.toEntryProcessorFn = toEntryProcessorFn;
+            this.isLocal = isLocal;
+        }
+
+        @Override
+        public boolean isCooperative() {
+            return false;
+        }
+
+        @Override
+        public boolean tryProcess() {
+            checkError();
+            return true;
+        }
+
+        @Override
+        protected boolean tryProcess(int ordinal, @Nonnull Object object) throws Exception {
+            checkError();
+            if (numConcurrentOps.incrementAndGet() >= MAX_PARALLEL_ASYNC_OPS) {
+                return false;
+            }
+            try {
+                T item = (T) object;
+                EntryProcessor<K, V> entryProcessor = toEntryProcessorFn.apply(item);
+                K key = toKeyFn.apply(item);
+                map.submitToKey(key, entryProcessor, callback);
+                return true;
+            } catch (HazelcastInstanceNotActiveException e) {
+                handleInstanceNotActive(instance, e, isLocal);
+                return false;
+            }
+        }
+
+        @Override
+        public boolean complete() {
+            return ensureAllWritten();
+        }
+
+        @Override
+        public boolean saveToSnapshot() {
+            return ensureAllWritten();
+        }
+
+        private boolean ensureAllWritten() {
+            checkError();
+            return numConcurrentOps.get() == 0;
```

Comment:
This two lines should be swapped: it can happen, that there are no errors, then last operation completes with error and the counter is 0 -> we'll miss the error.
Fix this also in `WriteKafkaP`.